File: Connect-Four-Game.py
'''
Assigning values to the grid
The grid will look like this:
  0,0 | 0,1 | 0,2 | 0,3 | 0,4 | 0,5 | 0,6
  1,0 | 1,1 | 1,2 | 1,3 | 1,4 | 1,5 | 1,6
  2,0 | 2,1 | 2,2 | 2,3 | 2,4 | 2,5 | 2,6
  3,0 | 3,1 | 3,2 | 3,3 | 3,4 | 3,5 | 3,6
  4,0 | 4,1 | 4,2 | 4,3 | 4,4 | 4,5 | 4,6
  5,0 | 5,1 | 5,2 | 5,3 | 5,4 | 5,5 | 5,6
'''
N, M = 6, 7
a_row = 4
grid = []
marks=['X','O']
#This function prints the grid of Connect Four Game as the game progresses
def print_grid():
    print("Player 1: X  vs  Player 2: O")
    print('--' + '---' * M + '--')
    for i in range(N):
        print(end='|  ')
        for j in range(M):
            print(grid[i][j], end='  ')
        print(end='|')
        print()
        print('--' + '---' * M + '--')

#This function checks the existence of any winning consecutive horizontal four 
def check_rows(grid):
    for row in range(N-1,-1,-1):
        for col in range(M-(a_row-1)):
            for i in range(col+1,col+a_row):
                if grid[row][col] == grid[row][i] and grid[row][col] in marks:
                    continue
                else:
                    break
            else:
                return True
    return False

#This function checks the existence of any winning consecutive vertical four 
def check_cols(grid):
    for col in range(M):
        for row in range(N-1,a_row-2,-1):
            for j in range(row-1,row-a_row,-1):
                if grid[row][col] == grid[j][col] and grid[row][col] in marks:
                    continue
                else:
                    break
            else:
                return True
    return False

#This function checks the existence of any winning consecutive diagonal four
def check_diags(grid):
    for row in range(N-1,a_row-2,-1):
        for col in range(M-1,a_row-2,-1):
            j = col-1
            for i in range(row-1,row-a_row,-1):
                if grid[row][col] == grid[i][j] and grid[row][col] in marks:
                    j -= 1
                    continue
                else:
                    break
            else:
                return True

    for row in range(N-1,a_row-2,-1):
        for col in range(0,M-(a_row-1)):
            j = col+1
            for i in range(row-1,row-a_row,-1):
                if grid[row][col] == grid[i][j] and grid[row][col] in marks:
                    j += 1
                    continue
                else:
                    break
            else:
                return True

    return False

#This function checks if row or column or diagonal is full with same characters
def check_win():
    if check_rows(grid) or check_cols(grid) or check_diags(grid):
        return True
    else:
        return False

#This function makes a temporary copy from grid and fill the spaces with specific mark
def list_try(mark):
    temp=[]
    for row in range(N):
        l = []
        for col in range(M):
            if grid[row][col] == '.':
                l.append(mark)
            else:
                l.append(grid[row][col])
        temp.append(l)
    return temp

#This function checks if the game has a tie state or not for the given mark
def check_tie_player(mark):
    temp = list_try(mark)
    if check_rows(temp) or check_cols(temp) or check_diags(temp):
        return False
    else:
        return True

#This function checks if there is a tie status
def check_tie():
    all_tie = True
    for mark in marks:
        if not check_tie_player(mark):
            all_tie = False
    return all_tie

#This function checks if given column has empty cell or not 
def check_empty_col_cell(i):
    index = None
    for j in range(N-1,-1,-1):
        if grid[j][i] =='.' :
            index = j
            break
    return index

#This function checks if given cell is empty or not 
def check_empty(i):
    if grid[0][i] == '.':
        return True
    else:
        return False

#This function checks if given position is valid or not 
def check_valid_column(i):
    if 0 <= i < M :
        return True
    else:
        return False

#This function sets a value to a cell
def set_cell(i, mark):
    row = check_empty_col_cell(i)
    grid[row][i] = mark


#This function clears the grid
def grid_clear():
    grid.clear()
    for i in range(N):    
        grid.append(['.']*M)


#MAIN FUNCTION
def play_game():
    print("Connect Four Game!")
    print("Welcome...")
    print("============================")
    player = 0
    while True:
        #Prints the grid
        print_grid()
        #Set mark value based on the player
        mark = 'X' if player == 0 else 'O'
        #Takes input from the user to fill in the grid
        print('Player %s' % mark)
        i = int(input('Enter the column index: '))
        while not check_valid_column(i) or not check_empty(i):
            i = int(input('Enter a valid column index: '))
        #Set the input position with the mark
        set_cell(i, mark)
        #Check if the state of the grid has a win state
        if check_win():
            #Prints the grid
            print_grid()
            print('Congrats, Player %s is won!' % mark)
            break
        
        #Check if the state of the grid has a tie state
        if check_tie():
            #Prints the grid
            print_grid()
            print("Woah! That's a tie!")
            break		
        #Player number changes after each turn
        player = 1 - player 


while True:
	grid_clear()
	play_game()
	c = input('Play Again [Y/N] ')
	if c not in 'yY':
		break

File: Connect-Four-Game.cpp
#include <bits/stdc++.h>
using namespace std;

const int N = 6;
const int M = 7;
const int a_row = 4;
const int n_players = 2;
char marks[n_players] = {'X', 'O'};
char grid[N][M];
char temp[N][M];
//This function prints the grid of Connect Four Game as the game progresses
void print_grid() {
	for (int i = 0; i< n_players; i++) {
        cout << "Player " << i+1 << ": " << marks[i] << "  ";
        if (i < n_players-1)
            cout << "vs  ";
	}
	cout << "\n";
	cout << "--";
    for (int i = 0; i < M; cout << "---", i++);
    cout << "--\n";
    for (int i = 0; i < N; i++) {
        cout << "|  ";
        for (int j = 0; j < M; j++)
            cout << grid[i][j] << "  ";
        cout << "|\n";
        cout << "--";
        for (int i = 0; i < M; cout << "---", i++);
        cout << "--\n";
    }
}

//This function checks if there is a win row
bool check_row(char grid[N][M]){
	for(int row=N-1;row>-1;row--){
		for(int col=0;col<M-(a_row-1);col++){
			for(int i = col+1;i<col+a_row;i++){
				if(grid[row][col]==grid[row][i] && (grid[row][col] == 'X' || grid[row][col] == 'O')){
					if(i==(col+a_row)-1){
						return true;
					}
					continue;	
				}
				else{
					break;
				}
			}
		}
	}
	return false;
}
//This function checks if there is a win column
bool check_col(char grid[N][M]){
	for(int col=0; col<M ; col++){
		for(int row = N-1 ; row > a_row-2 ; row--){
			for(int i = row-1 ; i > row-a_row ; i--){
				if (grid[row][col] == grid[i][col] && (grid[row][col] == 'X' || grid[row][col] == 'O')){
					if(i==(row-a_row)+1){
						return true;
					}
					continue;
				}
				else{
					break;
				}
			}
		}
	}
	return false;
}

//This function checks if there is a win diagonal
bool check_diag(char grid[N][M]){
	for(int row = N-1 ; row > a_row-2; row--){
		for(int col = M-1 ; col > a_row-2 ; col--){
			int j = col -1;
			for(int i = row-1 ; i > row-a_row;i--){
				if (grid[row][col] == grid[i][j] && (grid[row][col] == 'X' || grid[row][col] == 'O')){
					if(i==(row-a_row)+1){
						return true;
					}
					j -= 1;
					continue;
				}
				else{
					break;
				}
			}
		}
	}
	
	for(int row = N-1 ; row > a_row-2; row--){
		for(int col = 0 ; col < M-(a_row-1) ; col++){
			int j = col +1;
			for(int i = row-1 ; i > row-a_row;i--){
				if (grid[row][col] == grid[i][j] && (grid[row][col] == 'X' || grid[row][col] == 'O')){
					if(i==(row-a_row)+1){
						return true;
					}
					j += 1;
					continue;
				}
				else{
					break;
				}
			}
		}
	}
	return false;
}

//This function checks if the game has a win state or not
bool check_win() {
	if(check_row(grid) || check_col(grid) || check_diag(grid)){
		return true;
	}
	else{
		return false;
	}
}

//This function copy the main array(grid) to a temporary array(temp) and replace empty spaces with marks
void arr_try(char mark){
	for(int i = 0;i<N;i++){
		for(int j = 0;j<M;j++){
			if(grid[i][j]==' '){
				temp[i][j] = mark;
			}else{
				temp[i][j] = grid[i][j];
			}
		}		
	}
}

//This function checks if the game has a tie state or not for the given mark
bool check_tie_player(char mark) {
	arr_try(mark);
	if(check_row(temp)||check_col(temp)||check_diag(temp)){
		return false;
	}
	return true;
}
//This function checks if the game has a tie state or not
bool check_tie() {
    bool all_tie = true;
    for (int i = 0; i < n_players; i++)
        if (!check_tie_player(marks[i]))
            all_tie = false;
    return all_tie;
}
//This function checks if given cell is empty or not 
bool check_empty(int i) {
	if (grid[0][i] == ' '){
		return true;	
	}
	else{
		return false;
	}
}
//This function checks if given position is valid or not 
bool check_valid_column(int i) {
	if( 0 <= i && i < M){
		return true;
	}
	else{
		return false;
	}
}
//This function gets the first empty place in column
int get_empty(int i){
	for(int row = N-1 ; row>-1 ; row-- ){
		if (grid[row][i]==' '){
			return row;
		}
	}
}
//This function sets the given mark to the given column
void set_cell(int i, char mark) {
	int row = get_empty(i);
	grid[row][i]=mark;
}
//This function clears the game structures
void grid_clear() {
	for(int i = 0;i<N;i++){
		for(int j = 0;j<M;j++){
			grid[i][j] = ' ';
		}
	}
}
//This function reads a valid position input
void read_input(int &i) {
    cout << "Enter the column index: ";
    cin >> i;
    while (!check_valid_column(i) || !check_empty(i)) {
        cout << "Enter a valid column index: ";
        cin >> i;
    }
}
//MAIN FUNCTION
void play_game() {
    cout << "Connect Four Game!\n";
    cout << "Welcome...\n";
    cout << "============================\n";
    int player = 0;
    while (true) {
        //Prints the grid
        print_grid();
        //Read an input from the player
        cout << "Player " << marks[player] << " is playing now\n";
        int i;
		read_input(i);
        //Set the input position with the mark
        set_cell(i, marks[player]);
        //Check if the state of the grid has a win state
        if (check_win()) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "Congrats, Player " << marks[player] << " is won!\n";
            break;
        }
        //Check if the state of the grid has a tie state
        if (check_tie()) {
            //Prints the grid
            print_grid();
            //Announcement of the final statement
            cout << "Woah! That's a tie!\n";
            break;
        }
        //Player number changes after each turn
        player = (player + 1) % n_players;
    }
}
int main() {
    while (true) {
    	grid_clear();
    	play_game();
    	char c;
    	cout << "Play Again [Y/N] ";
    	cin >> c;
    	if (c != 'y' && c != 'Y')
    		break;
    }
}
